moveit_core/collision_detection/src/world.cpp

```diff
@@ -225,9 +225,14 @@ bool World::moveObject(const std::string& object_id, const Eigen::Isometry3d& tr
   if (transform.isApprox(Eigen::Isometry3d::Identity()))
     return true;  // object already at correct location
   ensureUnique(it->second);
-  for (size_t i = 0, n = it->second->shapes_.size(); i < n; ++i)
+  it->second->pose_ = transform * it->second->pose_;
+  for (Eigen::Isometry3d& shape_pose : it->second->shape_poses_)
```

Comment:
At the moment, all the transformations in `world.h` are global, so I stuck with that convention. Making the poses (and subframes) relative to the pose would be a step towards a scene graph, which I don't think should be in the scope of this PR.

Reply:
If you restrict the scope of this PR in this fashion, you don't need to explicitly store the pose of the collision object at all. Just transform all shape transforms accordingly.
This will avoid the initialization issues we see.

Reply:
But if the pose is explicitly stored the pose of an object becomes clearly visible in the data structure too (at the cost of an additional transform to store for each object).

I would also prefer to store relative transforms with this patch.
It's the main point of this refactoring after all and puts the additional transform to good use.

If you do not include it, because you would run into API problems or unexpected complexity,
please add an appropriate issue (best case assigned to yourself :sunglasses: ).

Reply:
Good that we agree on this. I was kind of confused that the feature of relative transforms wasn't utilized yet. 
However, we should discuss, maybe in tomorrow's maintainer meeting, whether we want to accept the additional matrix multiplication necessary to get absolute transforms (for each collision update).

Reply:
I could store the relative shape/subframe transforms and change all of the `getTransform` calls so they return `pose*transform`, but I suspect that the overhead of doing the multiplication every time will be significant. I would tend towards storing both global and local transforms, as suggested [here](https://github.com/ros-planning/moveit/pull/2002#issuecomment-624445849), but @v4hn said you were opposed at some point(?). Would be great to have a benchmark for this decision.

I know you also brought up the idea of lazy lookups, which sounds good but [this discussion of data structure options](https://github.com/ros-planning/moveit/issues/202) was inconclusive, and something like that would be yet a step further.

Maybe we can decide on something in the maintainer meeting.

Reply:
I just noticed that global and relative subframe poses are already stored in the `attached_body` since [this PR](https://github.com/ros-planning/moveit/pull/1757), so I would say that adding `relative` and `global` subframes to the world objects as well makes the most sense.

Reply:
I don't like that we attempt to mess up the whole frame handling even more instead of going for a clean scene graph interface.
If you start storing local and global transforms for subframes, you will need to do the same for the object shapes as well... Cleaning this up later is the same effort as adding it now...

Reply:
And I am worried adding a graph structure will blow this change out of proportion, since the scene graph implementation isn't any clearer, and it would add the need for lazy lookups.

Storing all the transformations as relative ones from the object's pose may be the best bet for now.